### PR TITLE
[CSL-1150] Merge cardano-sl-0.4-mid to master

### PR DESCRIPTION
**Pull request by georgeee**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -4,8 +4,15 @@ sudo: true
 matrix:
   include:
     - os: osx
-      osx_image: xcode8.3
+      osx_image: xcode7.3
```
Comment (Pastafarianist) [inline .travis.yml]: Is this intended?

File: src/smart-generator/Util.hs [hunk 2]
```diff
@@ -1,4 +1,6 @@
-module Util where
+module Util
+  ( module Util
```
Comment (Martoon-00) [inline src/smart-generator/Util.hs]: Wow, how does it work?

Comment (Martoon-00) [inline src/smart-generator/Util.hs]: I see know, it was `hlint` suggestion

Comment (georgeee) [inline src/smart-generator/Util.hs]: Basically same as no `(module Util)` at all, requested by linter

**Comment by gromakovsky** [comment]
Where are merge commits? What did you do?

**Comment by georgeee** [comment]
@gromakovsky fucking git. I switched branch after starting merge and it didn't preserve history. I merged with --ours now

File: src/node/Main.hs [hunk 3]
```diff
@@ -161,6 +164,96 @@ action peerHolder args@Args {..} transport = do
...
+        , npUseNTP = not noNTP
+        }
+
+gtSscParams :: Args -> VssKeyPair -> GtParams
+gtSscParams Args {..} vssSK =
+    GtParams
+    { gtpSscEnabled = True
+    , gtpVssKeyPair = vssSK
+    }
+=======
```
Comment (gromakovsky) [inline src/node/Main.hs]: It won't compile.

File: src/Pos/Client/Txp/History.hs [hunk 4]
```diff
@@ -40,32 +40,32 @@ import           Data.Tagged                  (Tagged (..))
...
 import           Pos.DB                       (MonadDB)
 import qualified Pos.DB.Block                 as DB
 import           Pos.DB.Error                 (DBError (..))
 import qualified Pos.DB.GState                as GS
 import           Pos.Slotting                 (MonadSlots)
 import           Pos.Ssc.Class                (SscHelpersClass)
-import           Pos.WorkMode.Class           (TxpExtra_TMP)
 #ifdef WITH_EXPLORER
-import           Pos.Explorer                 (eTxProcessTransaction)
+import           Pos.Explorer.Txp.Local       (eTxProcessTransaction)
```
Comment (Pastafarianist) [inline src/Pos/Client/Txp/History.hs]: Wait, I don't think this change was in https://github.com/input-output-hk/cardano-sl/pull/304. Why was it introduced?

File: .travis.yml (hunk 1 above)
Comment (georgeee) [inline .travis.yml]: I thout idea was to have both builds in CI. No problem to remove one later

File: src/Pos/Client/Txp/History.hs (hunk 4 above)
Comment (georgeee) [inline src/Pos/Client/Txp/History.hs]: To avoid circular dependency

File: src/Pos/Explorer/DB.hs [hunk 5]
```diff
@@ -29,6 +38,38 @@ getAddrHistory :: MonadDB m => Address -> m AddrHistory
...
                  gsGetBi . addrHistoryPrefix
 
+getAddrBalance :: MonadDB m => Address -> m (Maybe Coin)
+getAddrBalance = gsGetBi . addrBalancePrefix
+
+----------------------------------------------------------------------------
+-- Initialization
+----------------------------------------------------------------------------
+
+prepareExplorerDB :: (Ether.MonadReader' GenesisUtxo m, MonadDB m) => m ()
```
Comment (georgeee) [inline src/Pos/Explorer/DB.hs]: Earlier constraint was something like `WithNodeContext`, now more refined

File: .travis.yml (hunk 1 above)
Comment (georgeee) [inline .travis.yml]: CI fails on 7.3, removing.. :)

**Comment by georgeee** [comment]
One build hang on tets phase, not a big deal

**Pull request merged**